### PR TITLE
Address structure speed optimization

### PR DESCRIPTION
**Pull request by sfluhrer**
This is a speed optimization, where we keep the address structure in the format that the thash function will actually use them, rather than in the 8 word format (and converting them each time with the addr_to_bytes or compress_address function.  When one of the accessor functions (set_*) writes them, it writes the corresponding byte or bytes (a single byte for those fields who never will contain a value > 255).

In the best case (AVX2-accelerated Haraka), this has been seen to give a 16% speed-up; in other cases, the gains are more modest (2%-4%), but are still there.

I keep the address structure as 8 uint32_t's, even though the code doesn't use them that way; that's to limit the code churn required.

### CODE REVIEW

File: ref/address.c [hunk 1]
```diff
@@ -1,79 +1,111 @@
...
+ */
 void set_tree_addr(uint32_t addr[8], uint64_t tree)
 {
 #if (SPX_TREE_HEIGHT * (SPX_D - 1)) > 64
     #error Subtree addressing is currently limited to at most 2^64 trees
 #endif
-    addr[1] = 0;
-    addr[2] = (uint32_t) (tree >> 32);
-    addr[3] = (uint32_t) tree;
+    ull_to_bytes(&((unsigned char *)addr)[SPX_OFFSET_TREE], 8, tree );
```
Comment (bwesterb) [inline ref/address.c:23]: Not using the generic `ull_to_bytes`, but a new `u64_to_bytes` would be faster.

**Comment by bwesterb** [comment]
I like this.  This still breaks on big endian machines, right?  (Didn't check.)

File: ref/address.c (hunk 1 above)
Comment (sfluhrer) [reply ref/address.c:23]: True; however I don't believe performance is critical right here; set_tree_addr is not called that often (hence it doesn't make that much sense to implement a new u64_to_bytes function just for this)

**Comment by sfluhrer** [comment]
I don't think it breaks on bigendian machines; I'm pretty careful that way, however I haven't actually tested it

File: ref/address.c [hunk 2]
```diff
@@ -1,79 +1,111 @@
...
 }
 
+/*
+ * Specify the distance from the left edge of the node in the Merkle/FORS tree
+ * (the tree index)
+ */
 void set_tree_index(uint32_t addr[8], uint32_t tree_index)
 {
-    addr[7] = tree_index;
+    ull_to_bytes(&((unsigned char *)addr)[SPX_OFFSET_TREE_INDEX], 4, tree_index );
```
Comment (bwesterb) [inline ref/address.c:110]: u32_to_bytes was merged recently

**Comment by bwesterb** [comment]
> I don't think it breaks on bigendian machines; I'm pretty careful that way, however I haven't actually tested it

I double checked and where I thought there was an issue, I was mistaken.

**Comment by bwesterb** [comment]
This commit assumes w ≤ 256. Will that be a problem in the future?  We might add an `#error` because this will fail silently when we ever use w bigger than 256.

**Comment by sfluhrer** [comment]
Yes, as written, it assumes w <= 256.  When I wrote it I did not expect a larger w (not anticipating constant sum encoding).  If we do go with constant sum encoding (and w > 256; none of the parameter sets I tried actually use a larger w, however it was close at times), then we can add the one-line change to support those.

File: ref/address.c [hunk 3]
```diff
@@ -1,79 +1,111 @@
...
 
+/*
+ * Specify which Merkle tree within the level (the "tree address") we're working on
+ */
 void set_tree_addr(uint32_t addr[8], uint64_t tree)
 {
 #if (SPX_TREE_HEIGHT * (SPX_D - 1)) > 64
     #error Subtree addressing is currently limited to at most 2^64 trees
 #endif
-    addr[1] = 0;
```
Comment (bwesterb) [inline ref/address.c:25]: Was this one redundant to begin with?

**Comment by sfluhrer** [comment]
The check in set_tree_addr was there in the original.  The ADDR structure was designed to handle the general case; however, when the code was written, they noted that, if they assumed h-h/d <= 64, that is, the hypertree height minus one merkle tree height was no more than 64, that assumption simplified the code a lot (as the "tree addr" would fit in 64 bits), and so they made that assumption (with the explicit #if check to be certain).  I just kept that assumption (and my tooling never suggested a parameter set where that assumption was not true)

**Comment by bwesterb** [comment]
> The check in set_tree_addr was there in the original. The ADDR structure was designed to handle the general case; however, when the code was written, they noted that, if they assumed h-h/d <= 64, that is, the hypertree height minus one merkle tree height was no more than 64, that assumption simplified the code a lot (as the "tree addr" would fit in 64 bits), and so they made that assumption (with the explicit #if check to be certain). I just kept that assumption (and my tooling never suggested a parameter set where that assumption was not true)

I was not referring to the check on `SPX_D`, but to the line `addr[1] = 0`.

**Comment by bwesterb** [comment]
(I'm in favour or merging this, btw.)

File: ref/address.c (hunk 3 above)
Comment (sfluhrer) [reply ref/address.c:25]: The addr structure is always initialized to zero, hence explicitly setting it to zero is redundant (although, as I mentioned elsewhere, we are not time critical here)

File: ref/address.c (hunk 2 above)
Comment (sfluhrer) [reply ref/address.c]: Made the change

**Comment by sfluhrer** [comment]
@bwesterb It's been a while, however I got the compile errors fixed, and also did your requested u32_to_bytes change.  What needs to be done to merge this?

**Review by bwesterb — APPROVED** [review]
I'm happy with it.  @joostrijneveld ?

**Comment by joostrijneveld** [comment]
> I'm happy with it. @joostrijneveld ?

Sounds like this has been seen by at least four eyes ;)

LGTM :+1:

**Comment by bwesterb** [comment]
> > I'm happy with it. @joostrijneveld ?
> 
> 
> 
> Sounds like this has been seen by at least four eyes ;)
> 
> 
> 
> LGTM :+1: 

@joostrijneveld 

I don't seem to have write access. Could you merge it?

**Comment by joostrijneveld** [comment]
> I don't seem to have write access. Could you merge it?

Ah! I was under the impression that team members would be able to commit (and thus merge) to all team repositories. I think I gave all team members write permission now; let me know if it works!

**Comment by sfluhrer** [comment]
Thanks!

Now, I’ll get started on putting the finishing touches on my other optimization idea, which (according to my tests) speeds up the SHA256/AVX by 15% and SHAKE256/AVX and Haraka/AES_NI by 10% (on top of the improvement we just did).

In case you’re wondering, with the current x4/x8 implementations, they compute 4 or 8 of the FORS trees in parallel; however, if the number of FORS trees doesn’t happen to be a multiple of 4 or 8, the last pass will not do anything useful with some of the parallel channels.  What my changes do is change the order we compute the PRF/F/H functions, so that we process 4 or 8 nodes of a single FORS tree at once (and since the number of nodes in a FORS tree is always a multiple of 4 or 8, except for the top few levels, we don’t have wasted computation.  Similarly, we generate 4 or 8 WOTS public keys at once (rather than working on them individually, and so waste channels because WLEN is not a multiple of 4 or 8).

**Comment by bwesterb** [comment]
> > I don't seem to have write access. Could you merge it?
> 
> Ah! I was under the impression that team members would be able to commit (and thus merge) to all team repositories. I think I gave all team members write permission now; let me know if it works (or doesn't)!

It works now, thanks!

**Comment by bwesterb** [comment]
> Thanks! Now, I’ll get started on putting the finishing touches on my other optimization idea,

Ok, great!  Looking forward to them.

**Pull request merged**